xaitk_saliency/utils/masking.py

```diff
@@ -1,33 +1,179 @@
+    if ref_mat.ndim > 2:
+        s = (..., None)  # add channel axis for multiplication
+    # The Batch Operation -> Bulk apply mask matrices to input image matrix to
+    # generate occluded images.
+    occ_img_mats = np.ndarray([len(masks), *ref_mat.shape], dtype=ref_mat.dtype)
+    if fill is not None:
+        masks_sview = masks[s]
+        np.add(
+            (masks_sview * ref_mat),
+            ((UINT8_ONE - masks_sview) * fill),
+            out=occ_img_mats, casting="unsafe"
+        )
+    else:
+        np.multiply(masks[s], ref_mat, out=occ_img_mats, casting="unsafe")
+    return [
+        PIL.Image.fromarray(img_mat, mode=ref_mode)
+        for img_mat in occ_img_mats
+    ]
+
+
+def occlude_image_streaming(
+    ref_image: PIL.Image.Image,
+    masks: Iterable[np.ndarray],
+    fill: Optional[Union[int, Sequence[int]]] = None,
+    threads: Optional[int] = None,
+) -> Generator[PIL.Image.Image, None, None]:
+    """
+    Apply a number of input occlusion masks to the given reference image,
+    producing a list of images equivalent in length, and parallel in order, to
+    the input masks.
+
+    We expect the "mask" matrices and the image to be the same height and
+    width, and for the mask matrix values to be in the [0, 1] range.
+    In the mask matrix, values closer to 1 correspond to regions of the image
+    that should *NOT* be occluded.
+    E.g. a 0 in the mask will translate to "blacking out" the corresponding
+    location in the source image.
+
+    We optionally take in a fill-color value as input. This may be either a
+    scalar or per-channel sequence. When no fill is passed, black is used
+    (default absence of color).
+
+    Images output will mirror the input image format. As such, the `fill` value
+    passed must be compatible with the input image channels for broadcasting.
+    For example, a single channel input will not be able to be broadcast
+    against a multi-channel `fill` input. A ValueError will be raised by the
+    underlying numpy call in such cases.
+
+    Assumptions:
+      * Mask input is per-pixel. Does not accept per-channel masks.
+      * Fill value input is in an applicable value range supported by the input
+        image format, which is mirrored in output images.
+
+    :param ref_image: Original base image
+    :param masks: Mask images in the [N, Height, Weight] shape format.
+    :param fill: Optional fill color for the occluded regions as a single uint8
+        value or a per-channel sequence.
+    :param threads: Optional number of threads to use for parallelism when set
+        to a positive integer. If 0, a negative value, or `None`, work will be
+        performed on the main-thread in-line.
+
+    :raises ValueError: One or more input masks in the input iterable did not
+        match shape of the input reference image.
+
+    :return: List of masked images in PIL Image form.
+    """
+    img_shape = ref_image.size[::-1]  # into [h,w] format.
```

Comment:
This might also need to be changed for consistency purposes, based on the feedback above.
```
input_size = (ref_image.height, ref_image.width)
```

Reply:
Sure.